{"expand":"renderedFields,names,schema,operations,editmeta,changelog,versionedRepresentations","id":"12403677","self":"https://issues.apache.org/jira/rest/api/2/issue/12403677","key":"OPENJPA-715","fields":{"fixVersions":[{"self":"https://issues.apache.org/jira/rest/api/2/version/12313301","id":"12313301","description":"","name":"1.0.4","archived":false,"released":true,"releaseDate":"2010-09-03"},{"self":"https://issues.apache.org/jira/rest/api/2/version/12313328","id":"12313328","description":"","name":"1.2.1","archived":false,"released":true,"releaseDate":"2009-03-18"},{"self":"https://issues.apache.org/jira/rest/api/2/version/12313326","id":"12313326","description":"","name":"1.3.0","archived":false,"released":false}],"resolution":{"self":"https://issues.apache.org/jira/rest/api/2/resolution/1","id":"1","description":"A fix for this issue is checked into the tree and tested.","name":"Fixed"},"customfield_12312322":null,"customfield_12312323":null,"customfield_12312320":null,"customfield_12310420":"161025","customfield_12312321":null,"customfield_12312328":null,"customfield_12312329":null,"customfield_12312326":null,"customfield_12310300":null,"customfield_12312327":null,"customfield_12312324":null,"customfield_12312720":null,"customfield_12312325":null,"lastViewed":null,"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/3","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/major.svg","name":"Major","id":"3"},"labels":[],"customfield_12312333":null,"customfield_12312334":null,"customfield_12313422":"false","customfield_12312331":null,"customfield_12312332":null,"aggregatetimeoriginalestimate":null,"timeestimate":null,"customfield_12312330":null,"versions":[{"self":"https://issues.apache.org/jira/rest/api/2/version/12312341","id":"12312341","description":"","name":"1.0.0","archived":false,"released":true,"releaseDate":"2007-08-28"},{"self":"https://issues.apache.org/jira/rest/api/2/version/12312344","id":"12312344","description":"","name":"1.1.0","archived":false,"released":true,"releaseDate":"2008-05-22"},{"self":"https://issues.apache.org/jira/rest/api/2/version/12313102","id":"12313102","description":"","name":"1.2.0","archived":false,"released":true,"releaseDate":"2008-08-12"}],"customfield_12311120":null,"customfield_12313826":null,"customfield_12312339":null,"issuelinks":[],"customfield_12313825":null,"assignee":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=faywang","name":"faywang","key":"faywang","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Fay Wang","active":true,"timeZone":"Etc/UTC"},"customfield_12312337":null,"customfield_12313823":null,"customfield_12312338":null,"customfield_12311920":null,"customfield_12313822":null,"customfield_12312335":null,"customfield_12313821":null,"customfield_12312336":null,"customfield_12313820":null,"status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are not closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Done"}},"components":[],"customfield_12312026":null,"customfield_12312023":null,"customfield_12312024":null,"aggregatetimeestimate":null,"customfield_12312022":null,"customfield_12310921":null,"customfield_12310920":"202682","customfield_12312823":null,"creator":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ekinsokmen","name":"ekinsokmen","key":"ekinsokmen","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ekin Sokmen","active":true,"timeZone":"Etc/UTC"},"subtasks":[],"reporter":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ekinsokmen","name":"ekinsokmen","key":"ekinsokmen","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ekin Sokmen","active":true,"timeZone":"Etc/UTC"},"aggregateprogress":{"progress":0,"total":0},"customfield_12313520":null,"customfield_12310250":null,"progress":{"progress":0,"total":0},"customfield_12313924":null,"votes":{"self":"https://issues.apache.org/jira/rest/api/2/issue/OPENJPA-715/votes","votes":0,"hasVoted":false},"worklog":{"startAt":0,"maxResults":20,"total":0,"worklogs":[]},"customfield_12313920":null,"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/secure/viewavatar?size=xsmall&avatarId=21133&avatarType=issuetype","name":"Bug","subtask":false,"avatarId":21133},"timespent":null,"customfield_12314020":"{summaryBean=com.atlassian.jira.plugin.devstatus.rest.SummaryBean@36b04cbe[summary={pullrequest=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@652df717[overall=PullRequestOverallBean{stateCount=0, state='OPEN', details=PullRequestOverallDetails{openCount=0, mergedCount=0, declinedCount=0}},byInstanceType={}], build=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@717ec3ab[overall=com.atlassian.jira.plugin.devstatus.summary.beans.BuildOverallBean@7be107dc[failedBuildCount=0,successfulBuildCount=0,unknownBuildCount=0,count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], review=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@6e30c6a3[overall=com.atlassian.jira.plugin.devstatus.summary.beans.ReviewsOverallBean@5687369d[stateCount=0,state=<null>,dueDate=<null>,overDue=false,count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], deployment-environment=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@51e3d47d[overall=com.atlassian.jira.plugin.devstatus.summary.beans.DeploymentOverallBean@22cff67a[topEnvironments=[],showProjects=false,successfulCount=0,count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], repository=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@4a22295e[overall=com.atlassian.jira.plugin.devstatus.summary.beans.CommitOverallBean@25fedd59[count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], branch=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@6a71a5a7[overall=com.atlassian.jira.plugin.devstatus.summary.beans.BranchOverallBean@3750e2b5[count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}]},errors=[],configErrors=[]], devSummaryJson={\"cachedValue\":{\"errors\":[],\"configErrors\":[],\"summary\":{\"pullrequest\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"stateCount\":0,\"state\":\"OPEN\",\"details\":{\"openCount\":0,\"mergedCount\":0,\"declinedCount\":0,\"total\":0},\"open\":true},\"byInstanceType\":{}},\"build\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"failedBuildCount\":0,\"successfulBuildCount\":0,\"unknownBuildCount\":0},\"byInstanceType\":{}},\"review\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"stateCount\":0,\"state\":null,\"dueDate\":null,\"overDue\":false,\"completed\":false},\"byInstanceType\":{}},\"deployment-environment\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"topEnvironments\":[],\"showProjects\":false,\"successfulCount\":0},\"byInstanceType\":{}},\"repository\":{\"overall\":{\"count\":0,\"lastUpdated\":null},\"byInstanceType\":{}},\"branch\":{\"overall\":{\"count\":0,\"lastUpdated\":null},\"byInstanceType\":{}}}},\"isStale\":false}}","customfield_12314141":null,"customfield_12314140":null,"project":{"self":"https://issues.apache.org/jira/rest/api/2/project/12310351","id":"12310351","key":"OPENJPA","name":"OpenJPA","projectTypeKey":"software","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/projectavatar?pid=12310351&avatarId=10043","24x24":"https://issues.apache.org/jira/secure/projectavatar?size=small&pid=12310351&avatarId=10043","16x16":"https://issues.apache.org/jira/secure/projectavatar?size=xsmall&pid=12310351&avatarId=10043","32x32":"https://issues.apache.org/jira/secure/projectavatar?size=medium&pid=12310351&avatarId=10043"},"projectCategory":{"self":"https://issues.apache.org/jira/rest/api/2/projectCategory/10252","id":"10252","description":"","name":"OpenJPA"}},"aggregatetimespent":null,"customfield_12312520":null,"customfield_12314422":null,"customfield_12314421":null,"customfield_12314146":null,"customfield_12314420":null,"customfield_12314145":null,"customfield_12314144":null,"customfield_12314143":null,"resolutiondate":"2008-09-18T15:41:09.062+0000","workratio":-1,"customfield_12312923":null,"customfield_12312920":null,"customfield_12312921":null,"watches":{"self":"https://issues.apache.org/jira/rest/api/2/issue/OPENJPA-715/watchers","watchCount":1,"isWatching":false},"created":"2008-09-04T13:07:03.212+0000","updated":"2010-03-09T18:32:55.773+0000","timeoriginalestimate":null,"description":"While merging object A in following relationship [ A (*-*) B (1-*) C ] we get the following error. The test case is attached and I tested it with openjpa 1.0.0, 1.1.0 and 1.2.0 official releases with same error. The main problem is the error does not occur if there are just 1 or 2 C objects attached to B. If we add more C to B the probability of getting the error increases (sad but true). You can see a for loop in the test case and a constant named \"MAGICAL_NUMBER\". If we set the magical number to 3 or less than we don't get this error and all objects are persisted just fine. But if we increase it to 50 (to be sure) we get the error below. \n\nThere is also strange a workaround on line 63 of the test case. If we uncomment line 63 and just get the IDs of new C objects just after merge(A) but before commit () we can display proper IDs and the operation will be committed successfully. Otherwise the IDs of new C objects stay \"0\"  (again sad but true :).  Maybe this hint can help on understanding and solving the issue.\n\nThe test case is created depending on a real life scenario. We get the same error with derby and mysql.\n\nHere is the error after running the test case:\nTests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.433 sec <<< FAILURE!\ntestChainEntities(org.apache.openjpa.persistence.relations.TestChainEntities)  Time elapsed: 3.362 sec  <<< ERROR!\n<openjpa-1.1.0-r422266:659716 fatal store error> org.apache.openjpa.persistence.RollbackException: The transaction has been rolled back.  See the nested exceptions for details on the errors that occurred.\n        at org.apache.openjpa.persistence.EntityManagerImpl.commit(EntityManagerImpl.java:523)\n        at org.apache.openjpa.persistence.relations.TestChainEntities.chainUpdate(TestChainEntities.java:64)\n        at org.apache.openjpa.persistence.relations.TestChainEntities.testChainEntities(TestChainEntities.java:32)\n        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)\n        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)\n        at java.lang.reflect.Method.invoke(Method.java:585)\n        at junit.framework.TestCase.runTest(TestCase.java:154)\n        at junit.framework.TestCase.runBare(TestCase.java:127)\n        at junit.framework.TestResult$1.protect(TestResult.java:106)\n        at junit.framework.TestResult.runProtected(TestResult.java:124)\n        at junit.framework.TestResult.run(TestResult.java:109)\n        at junit.framework.TestCase.run(TestCase.java:118)\n        at org.apache.openjpa.persistence.test.PersistenceTestCase.run(PersistenceTestCase.java:122)\n        at junit.framework.TestSuite.runTest(TestSuite.java:208)\n        at junit.framework.TestSuite.run(TestSuite.java:203)\n        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)\n        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)\n        at java.lang.reflect.Method.invoke(Method.java:585)\n        at org.apache.maven.surefire.junit.JUnitTestSet.execute(JUnitTestSet.java:213)\n        at org.apache.maven.surefire.suite.AbstractDirectoryTestSuite.executeTestSet(AbstractDirectoryTestSuite.java:140)\n        at org.apache.maven.surefire.suite.AbstractDirectoryTestSuite.execute(AbstractDirectoryTestSuite.java:127)\n        at org.apache.maven.surefire.Surefire.run(Surefire.java:177)\n        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)\n        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)\n        at java.lang.reflect.Method.invoke(Method.java:585)\n        at org.apache.maven.surefire.booter.SurefireBooter.runSuitesInProcess(SurefireBooter.java:334)\n        at org.apache.maven.surefire.booter.SurefireBooter.main(SurefireBooter.java:980)\nCaused by: <openjpa-1.1.0-r422266:659716 fatal general error> org.apache.openjpa.persistence.PersistenceException: The transaction has been rolled back.  See the nested exceptions for details on the errors that occurred.\n        at org.apache.openjpa.kernel.BrokerImpl.newFlushException(BrokerImpl.java:2160)\n        at org.apache.openjpa.kernel.BrokerImpl.flush(BrokerImpl.java:2007)\n        at org.apache.openjpa.kernel.BrokerImpl.flushSafe(BrokerImpl.java:1905)\n        at org.apache.openjpa.kernel.BrokerImpl.beforeCompletion(BrokerImpl.java:1823)\n        at org.apache.openjpa.kernel.LocalManagedRuntime.commit(LocalManagedRuntime.java:81)\n        at org.apache.openjpa.kernel.BrokerImpl.commit(BrokerImpl.java:1347)\n        at org.apache.openjpa.kernel.DelegatingBroker.commit(DelegatingBroker.java:877)\n        at org.apache.openjpa.persistence.EntityManagerImpl.commit(EntityManagerImpl.java:512)\n        ... 29 more\nCaused by: <openjpa-1.1.0-r422266:659716 nonfatal general error> org.apache.openjpa.persistence.PersistenceException: The statement was aborted because it would have caused a duplicate key value in a unique or primary key constraint or unique index identified by 'SQL080903042502080' defined on 'CHAINENTITYC'. {prepstmnt 18739556 INSERT INTO ChainEntityC (cId, chainEntityBId, name, optLock, CHAINENTITYB_BID) VALUES (?, ?, ?, ?, ?) [params=(long) 0, (long) 0, (String) Test_C_48, (int) 1, (long) 3651]} [code=20000, state=23505]\nFailedObject: org.apache.openjpa.persistence.relations.ChainEntityC@f6f1b6\n        at org.apache.openjpa.jdbc.sql.SQLExceptions.narrow(SQLExceptions.java:146)\n        at org.apache.openjpa.jdbc.sql.DBDictionary.newStoreException(DBDictionary.java:4150)\n        at org.apache.openjpa.jdbc.sql.SQLExceptions.getStore(SQLExceptions.java:102)\n        at org.apache.openjpa.jdbc.sql.SQLExceptions.getStore(SQLExceptions.java:72)\n        at org.apache.openjpa.jdbc.kernel.PreparedStatementManagerImpl.flushAndUpdate(PreparedStatementManagerImpl.java:131)\n        at org.apache.openjpa.jdbc.kernel.BatchingPreparedStatementManagerImpl.flushAndUpdate(BatchingPreparedStatementManagerImpl.java:82)\n        at org.apache.openjpa.jdbc.kernel.PreparedStatementManagerImpl.flushInternal(PreparedStatementManagerImpl.java:89)\n        at org.apache.openjpa.jdbc.kernel.PreparedStatementManagerImpl.flush(PreparedStatementManagerImpl.java:72)\n        at org.apache.openjpa.jdbc.kernel.ConstraintUpdateManager.flush(ConstraintUpdateManager.java:543)\n        at org.apache.openjpa.jdbc.kernel.ConstraintUpdateManager.flush(ConstraintUpdateManager.java:105)\n        at org.apache.openjpa.jdbc.kernel.BatchingConstraintUpdateManager.flush(BatchingConstraintUpdateManager.java:56)\n        at org.apache.openjpa.jdbc.kernel.AbstractUpdateManager.flush(AbstractUpdateManager.java:89)\n        at org.apache.openjpa.jdbc.kernel.AbstractUpdateManager.flush(AbstractUpdateManager.java:72)\n        at org.apache.openjpa.jdbc.kernel.JDBCStoreManager.flush(JDBCStoreManager.java:549)\n        at org.apache.openjpa.kernel.DelegatingStoreManager.flush(DelegatingStoreManager.java:130)\n        ... 36 more\nCaused by: org.apache.openjpa.lib.jdbc.ReportingSQLException: The statement was aborted because it would have caused a duplicate key value in a unique or primary key constraint or unique index identified by 'SQL080903042502080' defined on 'CHAINENTITYC'. {prepstmnt 18739556 INSERT INTO ChainEntityC (cId, chainEntityBId, name, optLock, CHAINENTITYB_BID) VALUES (?, ?, ?, ?, ?) [params=(long) 0, (long) 0, (String) Test_C_48, (int) 1, (long) 3651]} [code=20000, state=23505]\n        at org.apache.openjpa.lib.jdbc.LoggingConnectionDecorator.wrap(LoggingConnectionDecorator.java:192)\n        at org.apache.openjpa.lib.jdbc.LoggingConnectionDecorator.access$700(LoggingConnectionDecorator.java:57)\n        at org.apache.openjpa.lib.jdbc.LoggingConnectionDecorator$LoggingConnection$LoggingPreparedStatement.executeUpdate(LoggingConnectionDecorator.java:866)\n        at org.apache.openjpa.lib.jdbc.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:269)\n        at org.apache.openjpa.jdbc.kernel.JDBCStoreManager$CancelPreparedStatement.executeUpdate(JDBCStoreManager.java:1398)\n        at org.apache.openjpa.jdbc.kernel.PreparedStatementManagerImpl.executeUpdate(PreparedStatementManagerImpl.java:151)\n        at org.apache.openjpa.jdbc.kernel.PreparedStatementManagerImpl.flushAndUpdate(PreparedStatementManagerImpl.java:120)\n        ... 46 more\n\n\nResults :\n\nTests in error:\n  testChainEntities(org.apache.openjpa.persistence.relations.TestChainEntities)\n\nTests run: 1, Failures: 0, Errors: 1, Skipped: 0\n\n[INFO] ------------------------------------------------------------------------\n[ERROR] BUILD FAILURE\n[INFO] ------------------------------------------------------------------------\n[INFO] There are test failures.\n","customfield_10010":null,"timetracking":{},"customfield_12314523":null,"customfield_12314127":null,"customfield_12314522":null,"customfield_12314126":null,"customfield_12314521":null,"customfield_12314125":null,"customfield_12314520":null,"customfield_12314124":null,"attachment":[{"self":"https://issues.apache.org/jira/rest/api/2/attachment/12389538","id":"12389538","filename":"openjpa-715.patch","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=faywang","name":"faywang","key":"faywang","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Fay Wang","active":true,"timeZone":"Etc/UTC"},"created":"2008-09-04T20:44:07.217+0000","size":1560,"mimeType":"text/plain","content":"https://issues.apache.org/jira/secure/attachment/12389538/openjpa-715.patch"},{"self":"https://issues.apache.org/jira/rest/api/2/attachment/12389497","id":"12389497","filename":"testcase_jira715.zip","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ekinsokmen","name":"ekinsokmen","key":"ekinsokmen","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ekin Sokmen","active":true,"timeZone":"Etc/UTC"},"created":"2008-09-04T13:08:47.217+0000","size":3250,"mimeType":"application/zip","content":"https://issues.apache.org/jira/secure/attachment/12389497/testcase_jira715.zip"}],"customfield_12312340":null,"customfield_12314123":null,"customfield_12312341":null,"customfield_12312220":null,"customfield_12314122":null,"customfield_12314121":null,"customfield_12310041":null,"customfield_12314120":null,"customfield_12314129":null,"customfield_12314524":null,"customfield_12314128":null,"summary":"OpenJpa does not generate IDs properly. \"duplicate key value in a unique or primary key constraint\" while merging object tree.","customfield_12314130":null,"customfield_12310291":null,"customfield_12310290":null,"customfield_12314138":null,"customfield_12314137":null,"environment":"Fedora 6\njava version \"1.5.0_11\"","customfield_12314136":null,"customfield_12314135":null,"customfield_12311020":null,"customfield_12314134":null,"duedate":null,"customfield_12314132":null,"customfield_12314131":null,"comment":{"comments":[{"self":"https://issues.apache.org/jira/rest/api/2/issue/12403677/comment/12628334","id":"12628334","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ekinsokmen","name":"ekinsokmen","key":"ekinsokmen","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ekin Sokmen","active":true,"timeZone":"Etc/UTC"},"body":"You can run test case by copying files to openjpa-persistence-jdbc module and running\nmvn test -Dtest=TestChainEntities","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ekinsokmen","name":"ekinsokmen","key":"ekinsokmen","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ekin Sokmen","active":true,"timeZone":"Etc/UTC"},"created":"2008-09-04T13:10:51.388+0000","updated":"2008-09-04T13:10:51.388+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12403677/comment/12628451","id":"12628451","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=faywang","name":"faywang","key":"faywang","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Fay Wang","active":true,"timeZone":"Etc/UTC"},"body":"The duplicate key problem is due to the fact that the primary key for ChainEntityC is only generated for the first two ChainEntityC objects in the collection. Starting from the third ChainEntityC object, the logic in openjpa is such that the object id will not be generated at all (it then assumes the default value of 0). That is why if the test caes has more than 3 ChainEntityC objects in the collection in ChainEntityB, we will see duplicate key error as more than one ChainEntityC has primary key = 0.  The attached patch fixes this problem. Any comment is mostly appreciated.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=faywang","name":"faywang","key":"faywang","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Fay Wang","active":true,"timeZone":"Etc/UTC"},"created":"2008-09-04T20:44:07.605+0000","updated":"2008-09-04T20:44:07.605+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12403677/comment/12628516","id":"12628516","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=faywang","name":"faywang","key":"faywang","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Fay Wang","active":true,"timeZone":"Etc/UTC"},"body":"As Kevin pointed out, sometimes the test case works fine with MAGICAL_NUMBER = 4 or 5 or 6 or .... In other words, the result is unpredictable. This is because during flush, the BrokerImpl will flush an un-ordered set of transactional objects. If the order of the flush is such that any C instance is flushed after A is flushed, that C instance will have valid object id. If all the C instances are flushed before A, then only the first 2 C instances will have valid object id.  \n\nSpecifically, right after A is done flushing, only the first 2 C instances have valid object id. If there are any C to be flushed in BrokerImpl, the beforeFlush of its state, PNewState, will be invoked, and assignObjectId will be called to generate valid object id. ","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=faywang","name":"faywang","key":"faywang","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Fay Wang","active":true,"timeZone":"Etc/UTC"},"created":"2008-09-04T23:23:10.319+0000","updated":"2008-09-04T23:23:10.319+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12403677/comment/12628664","id":"12628664","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ekinsokmen","name":"ekinsokmen","key":"ekinsokmen","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ekin Sokmen","active":true,"timeZone":"Etc/UTC"},"body":"I applied and tested the patch. The issue is solved. Thanks.\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ekinsokmen","name":"ekinsokmen","key":"ekinsokmen","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ekin Sokmen","active":true,"timeZone":"Etc/UTC"},"created":"2008-09-05T16:10:21.220+0000","updated":"2008-09-05T16:10:21.220+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12403677/comment/12629302","id":"12629302","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kwsutter","name":"kwsutter","key":"kwsutter","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kevin W. Sutter","active":true,"timeZone":"America/Chicago"},"body":"Resolved in 1.2.x and 1.3.0 (trunk).","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kwsutter","name":"kwsutter","key":"kwsutter","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kevin W. Sutter","active":true,"timeZone":"America/Chicago"},"created":"2008-09-08T21:37:54.138+0000","updated":"2008-09-08T21:37:54.138+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12403677/comment/12632250","id":"12632250","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kwsutter","name":"kwsutter","key":"kwsutter","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kevin W. Sutter","active":true,"timeZone":"America/Chicago"},"body":"Re-opening to migrate the fix back to the 1.0.x service stream.  A user requested this via private e-mail.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kwsutter","name":"kwsutter","key":"kwsutter","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kevin W. Sutter","active":true,"timeZone":"America/Chicago"},"created":"2008-09-18T15:29:13.944+0000","updated":"2008-09-18T15:29:13.944+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12403677/comment/12632254","id":"12632254","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kwsutter","name":"kwsutter","key":"kwsutter","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kevin W. Sutter","active":true,"timeZone":"America/Chicago"},"body":"Migrated change back to 1.0.x.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=kwsutter","name":"kwsutter","key":"kwsutter","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Kevin W. Sutter","active":true,"timeZone":"America/Chicago"},"created":"2008-09-18T15:41:09.043+0000","updated":"2008-09-18T15:41:09.043+0000"}],"maxResults":7,"total":7,"startAt":0},"customfield_12311820":"0|i0z2cf:","customfield_12314139":null}}

